File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        // return list
        List<List<Integer>> res = new ArrayList<>();
        
        // sort the given array
        Arrays.sort(nums);
        
        // first number
        int num1 = nums[0] - 1;
        
        for (int i = 0; i < nums.length; i++) {
            // check for dupulicate
            if (nums[i] != num1) {
                Map<Integer, Integer> map = new HashMap<>();
                
                // find numbers pair == -nums[i]
                int left = i + 1;
                int right = nums.length - 1;
                while (left < right) {
                    if (nums[left] + nums[right] == -nums[i]) {
                        map.put(nums[left], nums[right]);
                        left++;
                        right--;
                    }
                    else if (nums[left] + nums[right] < -nums[i])
                        left++;
                    else
                        right--;
                }
                
                // add to res
                if (!map.isEmpty()) {
                    for (Map.Entry<Integer, Integer> set : map.entrySet()) {
                        List<Integer> list = new ArrayList<>();
                        list.add(nums[i]);
                        list.add(set.getKey());
                        list.add(set.getValue());
                        res.add(list);
                    }
                }
            }
            // update num1
            num1 = nums[i];
        }
        return res;
    }
}

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        
        def twoSum(i):
            l = i + 1
            r = len(nums) - 1
            while l < r:
                sum = nums[l] + nums[r] + nums[i]
                if sum == 0:
                    res.append([nums[i], nums[l], nums[r]])
                    l += 1
                    r -= 1
                    while l < r and nums[l] == nums[l - 1]:
                        l += 1
                    while l < r and nums[r] == nums[r + 1]:
                        r -= 1
                elif sum > 0:
                    r -= 1
                else:
                    l += 1
            
        for i in range(len(nums)):
            if nums[i] > 0:
                break
            
            if i == 0 or nums[i] != nums[i - 1]:
                twoSum(i)
        
        return res